### ISSUE TITLE
SAM site stays after being sold during low-power

### ISSUE THREAD

**Issue by deniz1a** [OpenRA/OpenRA#7442]
![sam](https://cloud.githubusercontent.com/assets/6599106/6096234/4d6ab9c8-af8f-11e4-8c4e-e6b2adfd0db0.png)

It was targeting a plane while looking like that. I had to press sell multiple times before it was sold properly. This is in release-20141029. Might be related to #6076.

**Comment by pevers** [comment]
Reproducable:
1. Attack SAM site from enemy player that is powered down with AIRFORCE.
2. Sell SAM site on enemy side.

Doesn't matter if the player have low-power or that he just powered the SAM site down, effect is the same. 

![screenshot from 2015-02-23 20 44 27](https://cloud.githubusercontent.com/assets/3428649/6335774/2502dc7a-bb9e-11e4-9654-4bf1c0324a5a.png)

**Comment by penev92** [comment]
Thanks. Now that we have a reproduce-case, maybe someone can fix it :)

**Comment by jabbink** [comment]
This behaviour is induced by the AttackFollow Activity, which the SAM site has every tick, even when powered down.

Subsequently https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs#L74 will return early without going to the next activity (being sold).

I changed that line to return NextActivity, and selling a powered down, attacking SAM site worked fine. I am not sure what other side effects this has as I guess there is a reason why a disabled actor in AttackFollow mode should not do anything else.

**Comment by penev92** [comment]
First thing that comes to mind is an EMP-ed tank continuing to chase a target after the EMP effect wears off. But you are on a good path.
Consider also clearing the activity queue when a sell order is issued.

**Comment by obrakmann** [comment]
I cannot reproduce this on -playtest at all. Is this still an issue, or has it been fixed by accident?

**Comment by pevers** [comment]
@obrakmann Thats strange as I can still reproduce this on the latest bleed.

**Comment by pevers** [comment]
I think I found out why it isn't always repoducible. In 

https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/Sellable.cs#L46 

All the activities are cancelled before selling the building. But the animation will take some time to sell, so a new attack activity can be placed in the queue blocking the eventually sell (after the animation). I have proposed another fix in https://github.com/OpenRA/OpenRA/pull/7544,

**Comment by chrisforbes** [comment]
Selling should take the building lock. I think we shouldn't be queuing any
attack stuff if the building is locked.

- OR -

Is this not right? It would mean that capture in progress prevents firing
too. This would have balance implications.

**Comment by pevers** [comment]
@chrisforbes I think that would indeed mean that capturing a building would inmediatly disable attack which isn't preferred in my opinion

I outputted the queue activity in this situation:
```
cancelled all activity on sell
queue OpenRA.Mods.Common.Traits.AttackFollow+AttackActivity
queue OpenRA.Mods.Common.Activities.Sell
```

Preventing activities on disabled bulidings however will fix this bug and also improve a lot on efficiency I think. I'll try that out now.

**Comment by Phrohdoh** [comment]
Closed with #7544.
